pkg/reconcilermanager/controllers/rootsync_controller.go

```diff
@@ -779,52 +794,147 @@ func (r *RootSyncReconciler) validateRootSecret(ctx context.Context, rootSync *v
+	if err := r.client.List(ctx, &crbList, opts...); err != nil {
+		return nil, errors.Wrap(err, "listing ClusterRoleBindings")
+	}
+	for idx, crb := range crbList.Items {
+		roleRef := v1beta1.RootSyncRoleRef{
+			RoleRefBase: v1beta1.RoleRefBase{
+				Kind: crb.RoleRef.Kind,
+				Name: crb.RoleRef.Name,
+			},
+		}
+		currentRoleMap[roleRef] = &crbList.Items[idx]
 	}
+	return currentRoleMap, nil
+}
 
-	if roleConfig.Name == "" {
-		roleConfig.Name = "cluster-admin"
+func (r *RootSyncReconciler) cleanRBACBindings(ctx context.Context, reconcilerRef types.NamespacedName) error {
+	currentRefMap, err := r.listCurrentRoleRefs(ctx, reconcilerRef)
+	if err != nil {
+		return err
+	}
+	for _, binding := range currentRefMap {
+		if err := r.cleanup(ctx, binding); err != nil {
+			return errors.Wrap(err, "deleting RBAC Binding")
+		}
+	}
+	if err := r.deleteSharedClusterRoleBinding(ctx, RootSyncLegacyClusterRoleBindingName, reconcilerRef); err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "deleting legacy binding")
 	}
+	if err := r.deleteSharedClusterRoleBinding(ctx, RootSyncBaseClusterRoleBindingName, reconcilerRef); err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "deleting base binding")
+	}
+	return nil
+}
 
-	_, err := r.upsertClusterRoleBinding(ctx, reconcilerRef, roleConfig.Name)
-	return errors.Wrap(err, "upserting cluster role binding")
+func (r *RootSyncReconciler) configureClusterRoleBinding(ctx context.Context, reconcilerRef types.NamespacedName, roleRefs []v1beta1.RootSyncRoleRef) error {
+	currentRefMap, err := r.listCurrentRoleRefs(ctx, reconcilerRef)
+	if err != nil {
+		return err
+	}
+	if len(roleRefs) == 0 {
+		// Backwards compatible behavior to default to cluster-admin
+		if err := r.upsertSharedClusterRoleBinding(ctx, RootSyncLegacyClusterRoleBindingName, "cluster-admin", reconcilerRef); err != nil {
+			return err
+		}
+		// Clean up any RoleRefs created previously that are no longer declared
+		for _, binding := range currentRefMap {
+			if err := r.cleanup(ctx, binding); err != nil {
+				return errors.Wrap(err, "deleting RBAC Binding")
+			}
+		}
+		if err := r.deleteSharedClusterRoleBinding(ctx, RootSyncBaseClusterRoleBindingName, reconcilerRef); err != nil && !apierrors.IsNotFound(err) {
+			return errors.Wrap(err, "deleting base binding")
+		}
+		return nil
+	}
+	// Add the base ClusterRole for basic root reconciler functionality
+	if err := r.upsertSharedClusterRoleBinding(ctx, RootSyncBaseClusterRoleBindingName, RootSyncBaseClusterRoleName, reconcilerRef); err != nil {
+		return err
+	}
+	declaredRoleRefs := roleRefs
+	// Create RoleBindings which are declared but do not exist on the cluster
+	for _, roleRef := range declaredRoleRefs {
+		if _, ok := currentRefMap[roleRef]; !ok {
+			// we need to call a separate create method here for generateName
+			if _, err := r.createRBACBinding(ctx, reconcilerRef, roleRef.RoleRefBase, roleRef.Namespace); err != nil {
```

Comment:
Do we need to verify whether the referenced Role/ClusterRole exists, and fail earlier if it doesn't exist?

Reply:
I'm not sure it would add much value to verify that earlier on, as such condition would cause creation of the binding to error and surface on the RSync. Even if we did check for the existence of the Role/ClusterRole early on, it's always possible for race condition where another client deletes the Role/ClusterRole after our existence check.